pkgs/build-support/docker/default.nix

```diff
@@ -32,7 +32,33 @@ rec {
     inherit pkgs buildImage pullImage shadowSetup buildImageWithNixDb;
   };
 
-  pullImage = callPackage ./pull.nix {};
+  pullImage =
+    let
+      fixName = name: builtins.replaceStrings ["/" ":"] ["-" "-"] name;
+      ref = name: digest: tag: if digest != ""
+        then "docker://${name}@${digest}"
+        else "docker://${name}:${tag}";
+    in {
+      imageName, sha256,
+      # If ignored, image tag is then used to pull the image.
+      # To find the digest of an image, you can use skopeo:
+      # skopeo inspect docker://docker.io/nixos/nix:1.11 | jq -r '.Digest'
+      # sha256:20d9485b25ecfd89204e843a962c1bd70e9cc6858d65d7f5fadc340246e2116b
+      imageDigest ? "",
+      # If the image digest is specified, the tag is only used at
+      # Docker image loading (to set a tag), ie. it is not used to
+      # pull the image (both digest and tag are not supported by
+      # Skopeo yet).
```

Comment:
Does this mean we can’t fetch specific tags with this new implementation? Or does it mean you have to fetch the digest first and use that? What happens if the tag is set but the digest isn’t?

Reply:
We can still use image tags to pull images.
If the tag is set but the digest is not, the tag is used to pull the image.
But if you specify a digest and a tag, Skopeo ignores the tag to pull the image. In this case, the tag is just used at image creation.

Reply:
Hm, then it should really be `ref` (either a tag or a digest) and `finalImageName ? ref` (what the image is named). That way you don’t have to ignore user input silently.

Reply:
The tag is only ignored to pull the image if a specified image digest doesn't match to the specified image tag. In this case, we could ensure the digest matches the expected tag by using `skopeo inspect` before pulling the image.

I would like to avoid to have to encode the attribute type (digest or tag) in the reference string. For the user, it is more convenient to just have to set an attribute named `digest` or `tag` instead of having to put a `:` or a `@` in the image reference string. 
Moreover, the tag is always required (even if digest is specified) to create the image.

Another way would be to create a `pullImage` that only use digest and a `pullImageByTag` function that could take different type of arguments.
I'm not sure if we have to let - forever - the user pull images with a tag since they are not immutable.